pmlc/dialect/tile/transforms/expand_reshape.cc

```diff
@@ -0,0 +1,233 @@
+    while (pSrc >= 0 && srcDims[pSrc] == 1) {
+      --pSrc;
+      mulSrc = 1;
+    }
+    if (pSrc < 0) {
+      break;
+    }
+    while (pDst >= 0 && dstDims[pDst] == 1) {
+      --pDst;
+      mulDst = 1;
+    }
+    if (pDst < 0) {
+      break;
+    }
+    int64_t gcd = std::gcd(srcDims[pSrc], dstDims[pDst]);
+    assert(gcd > 1);
+    // The dim should be reversed
+    int idxId = numIdxs - ranges.size() - 1;
+    assert(idxId >= 0);
+    ranges.emplace_back(gcd);
+    srcExprs[pSrc] = srcExprs[pSrc] + getAffineConstantExpr(mulSrc, context) *
+                                          getAffineDimExpr(idxId, context);
+    dstExprs[pDst] = dstExprs[pDst] + getAffineConstantExpr(mulDst, context) *
+                                          getAffineDimExpr(idxId, context);
+    srcDims[pSrc] /= gcd;
+    dstDims[pDst] /= gcd;
+    mulSrc *= gcd;
+    mulDst *= gcd;
+  }
+
+  while (pSrc >= 0) {
+    if (srcDims[pSrc] > 1) {
+      unsigned idxId = numIdxs - ranges.size() - 1;
+      assert(idxId >= 0);
+      ranges.emplace_back(srcDims[pSrc]);
+      srcExprs[pSrc] = srcExprs[pSrc] + getAffineConstantExpr(mulSrc, context) *
+                                            getAffineDimExpr(idxId, context);
+    }
+    --pSrc;
+    mulSrc = 1;
+  }
+  while (pDst >= 0) {
+    if (dstDims[pDst] > 1) {
+      unsigned idxId = numIdxs - ranges.size() - 1;
+      assert(idxId >= 0);
+      ranges.emplace_back(dstDims[pDst]);
+      dstExprs[pDst] = dstExprs[pDst] + getAffineConstantExpr(mulDst, context) *
+                                            getAffineDimExpr(idxId, context);
+    }
+    --pDst;
+    mulDst = 1;
+  }
+
+  AffineMap srcMap = AffineMap::get(numIdxs, 0, srcExprs, context);
+  AffineMap sinkMap = AffineMap::get(numIdxs, 0, dstExprs, context);
+
+  // Build contraction for the reshape
+  auto elementType = src.getType().cast<RankedTensorType>().getElementType();
+  auto dstType = RankedTensorType::get(dstShape, elementType);
+  auto ident = tile::createIdentity(builder, builder.getUnknownLoc(),
+                                    elementType, AggregationKind::assign);
+  return builder.create<ContractionOp>(
+      builder.getUnknownLoc(),
+      /* resultType = */ dstType,
+      /* init = */ ident,
+      /* tensors = */ ArrayRef{src},
+      /* agg = */ util::AggregationKind::assign,
+      /* combo = */ util::CombinationKind::none,
+      /* sink = */ sinkMap,
+      /* srcs = */ ArrayRef{srcMap},
+      /* cons = */ IntegerSet::getEmptySet(0, 0, context),
+      /* name = */ "reshape");
+}
+
+struct ExpandReshapePass : public ExpandReshapeBase<ExpandReshapePass> {
+  void runOnFunction() final;
+
+  // Expand in-order reshape operation
+  void expandReshape(ReshapeOp reshapeOp);
+};
+
+// Expand in-order reshape operation
+void ExpandReshapePass::expandReshape(ReshapeOp reshapeOp) {
+  auto src = reshapeOp.tensor();
+  auto dst = reshapeOp.getResult();
+  auto dstShape = dst.getType().cast<RankedTensorType>().getShape();
```

Comment:
Is it guarantied that the destination type will be a `RankedTensorType`? I'm thinking of cases where we are casting to or from a scalar.

Reply:
I'm not sure. So I added a check now and see if there's any exception.